src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIProxy.cs

```diff
@@ -698,4 +576,260 @@ public SNIError GetLastError()
+
+            if (details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            if (!details.InferConnectionDetails() && details.IsBadDataSource)
+            {
+                return null;
+            }
+
+            return details;
+        }
+
+        private bool InferLocalServerName()
+        {
+            // If Server name is empty or localhost, then use "localhost"
+            if (string.IsNullOrEmpty(ServerName) || IsLocalHost(ServerName))
+            {
+                ServerName = ConnectionProtocol == DataSource.Protocol.Admin ?
+                    Environment.MachineName : DefaultHostName;
+            }
+            return true;
+        }
+
+        private bool InferConnectionDetails()
+        {
+            string[] tokensByCommaAndSlash = _dataSourceAfterTrimmingProtocol.Split(BackSlashSeparator, ',');
+            ServerName = tokensByCommaAndSlash[0];
+
+            int commaIndex = _dataSourceAfterTrimmingProtocol.IndexOf(',');
+
+            int backSlashIndex = _dataSourceAfterTrimmingProtocol.IndexOf(BackSlashSeparator);
+
+            // Check the parameters. The parameters are Comma separated in the Data Source. The parameter we really care about is the port
+            // If Comma exists, the try to get the port number
+            if (commaIndex > -1)
+            {
+                bool commaAfterSlash = commaIndex > backSlashIndex ? true : false;
+
+                string parameter = backSlashIndex > -1
+                        ? commaIndex > backSlashIndex ? tokensByCommaAndSlash[2].Trim() : tokensByCommaAndSlash[1].Trim()
+                        : tokensByCommaAndSlash[1].Trim();
+
+                // Bad Data Source like "server, "
+                if (string.IsNullOrEmpty(parameter))
+                {
+                    return ReportSNIError(SNIProviders.INVALID_PROV);
+                }
+
+                // For Tcp and Only Tcp are parameters allowed.
+                if (ConnectionProtocol == DataSource.Protocol.None)
+                {
+                    ConnectionProtocol = DataSource.Protocol.TCP;
+                }
+                else if (ConnectionProtocol != DataSource.Protocol.TCP)
+                {
+                    // Parameter has been specified for non-TCP protocol. This is not allowed.
+                    return ReportSNIError(SNIProviders.INVALID_PROV);
+                }
+
+                int port;
+                if (!int.TryParse(parameter, out port))
+                {
+                    return ReportSNIError(SNIProviders.TCP_PROV);
+                }
+
+                Port = port;
+            }
+
+            // Instance Name Handling. Only if we found a '\' and we did not find a port in the Data Source
+            if (backSlashIndex > -1 && Port == -1)
+            {
+                // This means that there will not be any part separated by comma. 
+                InstanceName = tokensByCommaAndSlash[1].Trim();
```

Comment:
I don't want to get any more Tcp Related validation logic in SQL. 
Whether a port is appropriate or not, should be a validation in the TCP layer, and that's what is happening with the Socket APIs. 
The Socket APIs take a `int` port number and we should rely on the networking layer to do the right thing. 
Otherwise, all this validation logic will be scattered in Sql and Networking.